### ISSUE TITLE
accepting invalid tls cert

### ISSUE THREAD

**Issue by szbnwer** [luakit/luakit#325]
hi all! :)

i'm fresh with luakit, but i like it. til now, i only got 3 errors, but i only can't handle this one. so i'm developing a site (not mine) where's no valid cert installed, so i got 'Unacceptable TLS certificate' error. this is fine, but i can't find a way to make invalid cert accepted without hacking in the codebase, and i'm not familiar with it, nor c side lua programming or gtk (mostly i can read them) so i didn't dare to mess with codes.

i'm using currently the latest ubuntu package, but mostly i can compile whatever i really want.

so many thanks for the great app, and as well if you can give me any time-efficient way to accept the invalid cert.

**Comment by aidanholm** [comment]
Hey :)

You should compile luakit from the `develop` branch, as I'm pretty certain the latest ubuntu package is massively outdated and insecure. I'm about to make a full release in a few days, after which packages will eventually be updated. Unfortunately the latest version is backwards incompatible in some ways.

In the updated version, the error page for invalid certificates has a button to ignore the security risk and access the website anyway, for the rest of the session. It's also easy to whitelist specific domains.

**Comment by szbnwer** [comment]
thx very much!

it wasn't that easy, but finally done and took all 3 issues :) (btw: other 2 was a facebook absolute positioned div from the right side of posts jumped in front of the posts when i scrolled down, but i killed that whole div from js console; and youtube volume button image was wrecked)

now i've got a 4th one here in this textarea, that blue shapes appears when clicking somewhere in the text, only after it already got the focus, but printscreen, clicking outside from the textarea and typing making it disappear, and i can draw on the shape by moving promt with the arrow buttons. i could even live with this, but if you can't reproduce this bug, then you can involve me into the research, but currently i've got no idea where could i start to research this, if you want it, then i can open a new issue for this (if new) but i couldn't say more about it, so it's possibly faster to copypaste than writing back for me...

thx again, and all the bests! - written from my brand new kawl browser :D

**Comment by aidanholm** [comment]
For the site-specific bugs, try setting your useragent to a different value. I have the following in my personal config:
```
globals.useragent = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/59.0.3071.86 Safari/537.36"
```
This just mimics a recent version of chrome; I update it when I remember to, and all the sites I use work flawlessly.

You're right, I had a look at the relevant code and it would actually be tricky to automatically whitelist sites---looks like the API is more complicated than I remember, sorry! The API isn't great at the moment; I've opened #327 about that.

The blue shapes appear in GitHub comment boxes? That's this WebKit bug: https://bugs.webkit.org/show_bug.cgi?id=155821. You can work around that with a user style that changes the CSS on the comment boxes.

Glad you're enjoying it :)

**Comment by szbnwer** [comment]
i think there's no much upon user agent, on server side i think its only for fingerprinting, and possibly for locking out from weird sites :D most likely some js can depend on that, but i think these are rare and mostly they can set different css rules that are browser dependent, but they are mostly just putted all together behind the same selectors...

i can live with the button for accept it every time, but anyway i've seen your new issue, cuz i subscribed to luakit github stuffs :)

that blue shape is kinda old then :D seems like none cares or they have no idea for it... :/ and yep, it have the same color as on the 2nd pic, and the same shape as the 1st (trapeziod when its fully appears)

most i can miss is some more comfort, like a menu for the basic tasks i can do, instead of hotkeys and a prompt, as this way i can't suggest it for old machines as lightweight goodness, because it's not really easy to use (ok, by the time i will be able to do anything i want, but i'm a senior :D ) and maybe a few tools i can have in firefox (the only problem with ff is that it's eating my ram...) like the web developer toolbar, facebook disconnect, tab suspender, youronlinechoices, greasemonkey, httpfox and umatrix; and pwmanager + formfiller; and fixed tabs (i have currently 16 of them, and i have like 100 others opened :D because developing learning and the like) - so i think i can get these by the time, as nothing is impossible after i made it from source once, and there are possible easy ways for many of them to get something like, but i will need mostly these for heavy use, and i still don't know how memory consuming it will be under pressure after moving from ff... if you have some instant hints, then many thx for them! :) (i already explored the resources that can be reached by following the links here and there, but i'm wondering if there is any collection for plugins or anything like out there; i've seen only a very few from the sources and docs there)

an another deep question inside me is the license (later in this paragraph, just it became a bit messy, sorry :D ), as i wanna put a browser engine in my hobby app sooner or later, but now i'm happy with a freestanding lightweight browser that is hackable via lua :) (my main set is ubuntu mate, bash, c, luajit, tekui (gui), turbo server; and i also know much other stuffs but not gtk for example, which i don't know why included, when there are no much in the window, and i think all that belongs to the page is for webkit, maybe the inspector can be gtk stuff... i'd like to move it into tekui first, but this is a very very big task, and then maybe replace c stuffs with ffi bindings or just get the point from the source where needed and build an own browser, but thats not just a very very big task, and this is not even a main goal for me... but i'd like to handle web contents from my app, and i don't really feel the need of rendering properly and running js, just get the point of functionality, and the identity with the version that i'm relying on, and then make completely alternative interfaces for whatever i want...) and i am willing to make things from the ground and i wanna own my tools so i need lightweight bases with as few dependencies as possible... so gpl seems good as it's like garlic against vampires (big corps that could wreck my 4k+ lines of plans that i wanna bring to the same fund - because they have team, money, and poisonous intentions like closed source (ok thats a hard point anyhow for them, but it's enough to take my features and make them shiney for the crowd, so they wont came here but there, and they can have closed source even as funds), like marketing (i wanna give it for the humanity for no cost), like overwhelm codebase and put shine on it instead of keep it lightweight, handleable and flexible, so they can kill my plans easily by a lotsa things)... on the other hand mit license would be good to give back anything to the origins, but that would turn into gpl after i'm using anything that is gpl licensed, so they have to change license in case of they would like accept anything from me - so best i can do now for protect my stuffs is to keep it private, and later i will give it to people who hate win and mac :D and when it is strong enough and knows enough, then i will make it public, but this is a very very long journey as i'm in lack of time and i have so much to do...

and anyway sorry for trashing the issues with carrying multiple topics parallel and can be mostly off topic :D but if you can give me an another place for continue the discussions, then right :)

and thx for all! :)

**Comment by aidanholm** [comment]
I know google serves different JS depending on the user agent (see https://github.com/aidanholm/luakit/issues/4), but if it works for you that's good. And yeah, fingerprinting is also a concern.

Luakit has the webkit inspector (the `:inspect` command), facebook disconnect can be replaced with the built-in adblocker and a suitable filter list (not included), there's a built-in greasemonkey-compatible user scripts manager (the `:userscripts` command), and there's a built-in formfiller.

We don't yet have a tab suspender, umatrix-like blocker, or fixed tabs. You can mimic fixed tabs by adding `w:new_tab()` calls in your config, so that they open automatically with every new session. httpfox seems to be a kind of network traffic analyzer; I think this is the job of external tools such as wireshark, although the webkit inspector also shows basic information on network traffic. I'm not sure what youronlinechoices does, apart from ensure a cookie is set? The cookies database is stored as an SQLite file, so you can always have a shell script manually edit that.

There aren't any other sources for plugins compatible with this release, as far as I am aware: any plugin found elsewhere would need to be updated.

tekUI looks really interesting! But unfortunately it'll be impossible to use WebKit with tekUI without GTK, unless you write and maintain your own WebKit port (a daunting task). You'd need to somehow embed a GTK widget within the tekUI widget hierarchy, then integrate the tekUI and GTK message loops, which may be impossible. A more viable alternative is to embed chromium via the CEF API, as you can use a tekui opengl drawing area widget (if it has one!).

**Comment by aidanholm** [comment]
@szbnwer could you try the fix in #329 and see if that fixes GitHub comments for you? Just put that CSS in a css file in `~/.local/share/luakit/styles/` and restart luakit, and it should work.

**Comment by szbnwer** [comment]
yep i've found the inspector, thats nice to have, but web developer toolbar is a different beast.. others are fine then, but where can i find the formfiller? and how can i reorder my opened tabs? :D

userscripts will be good for erase junk here and there on pages, but umatrix is still a good matery as thats a preventative blocker, instead of optimizer after the page is well known... can you give a hello world like stuff for the usage of userscripts? (or any link or whatever)

where to use this w:new_tab() ? in ~/.local/share/luakit/ there are no much stuffs to start to mess with intuitively

yep, i've used httpfox for analyzing https traffic, that can be achieved via wireshark, but thats tricky, and needed to do the trick for every site i wanna observe. i've used it to hack around an api (cpanel->cron->dehydrated let's encrypt client->php callback that logins with curl and installs the new cert x'D) but that can be okay, even using firefox for developing can be okay, as mostly facebook, gmail and the like are eating my ram there (4gib), so i can move with those here, but using 2 browsers can make mess around :D

and anyway i will give a try also for palemoon, but i think i won't abandon luakit because lua is my very favorit programming lamguage, but i've seen around webkit stuffs that there is c++ behind that, what i don't really like for tools i wanna go deeper with... and even a small thing with big dependencies aren't that small in my eyes, but this just means that i won't include luakit into my project, but i think i will use it as i will get more familiar with it...

i don't know how youronlinechoces works, but its increasing privacy on a lotsa places by a policy, its possible that it's using cookies..

tekui is really a very great 'matery', but it's kinda raw currently, and it's original developing is slow. but i won't find a better tool for my goals... it have nothing around gtk and opengl, it's an embedded project and minimalistic, but the most flexible i can get, and as it's small, readable and well shaped, i can customize anything around it...

i've just now seen, that webkit on linux depends on gtk..., i've thought before, that it's included because of the top and footer bars, the right click menu and the inspector, plus maybe a bunch of invisible features... so i think i won't integrate luakit in my project, but i'll share my tools between luakit and my app, and later i'll use a headless browser or a very own, but that will be a big pain :D i don't even care about daily usage, and fully featured stuff, i wanna know when a page changed (like rewritten, but content is a different thing...) and wanna communicate with servers from lua, possibly a libcurl is enough, but processing js would be also a very good point, my main goal is to automatize and customize interfaces...

#329 seems like did the trick :) i've made a github.css for it anyway, but no idea how luakit operates there...

and i've just found that ctrl+c is not working but clipboard is fine... tekui have clipboard issues (sys->tekui and tekui->another instance works fine, but tekui->sys is not working) maybe because i've installed it under my home folder instead of syswide, just as i did with luakit, but it became a bit messy to be able to start from the sources dir, as previously i've modified the PREFIX for make everywhere, and all those were installed under /usr/local/... gone to a local directory, but here ~/.local, /etc and whatever are included, so i have to take a deeper visit around the make process, as currently i need this for start it :D
cd $BASEDIR/source/luakit
export LD_LIBRARY_PATH=$BASEDIR/install/lib/
export PKG_CONFIG_PATH=$BASEDIR/install/lib/pkgconfig/
$BASEDIR/source/luakit/luakit
where the install dir is the prefix for luajit and the rest of my stuffs, and it's only the /usr/local, but no /etc and /home are included... i will find my way for this anyway, maybe it's not even necessary to make it dependent on my local installs

and one more point is that i miss the spell checker

so mostly these were informative, what i'd like to get is a start to use userscript, and a place for w:new_tab()

and once again many thx for your time and efforts! :) i hope i can be this much useful for you by the time :)

**Comment by aidanholm** [comment]
Do you mean this web developer toolbar? https://addons.mozilla.org/en-US/firefox/addon/web-developer/ we currently don't have that, but it'd be possible to write something like it. See luakit://help/doc/modules/formfiller.html for info on the formfiller. Tabs can be reordered with the < and > keys.

There's the `:userscriptinstall` and `:userscripts` commands. Browse to a userscript (the URL has to end in `.user.js`) and `:userscriptinstall` will let you install it. Umatrix-like functionality would be nice, but WebKit is missing a few needed APIs.

To automatically open a set of URIs on startup, edit `config/rc.lua`. Add the following line just before `window.new(uris)`. Should be around line 192, and look like this:
```lua
...
else
    -- Or open new window
    uris = lousy.util.table.join({"uri1", "uri2"}, uris)
    window.new(uris)
end
...
```

An alternative to wireshark is writing a small proxy in lua that can automatically hack around APIs, based on the URLs.

WebKitGTK isn't the only webkit port for linux; there's also a QT port. The GTK port uses GTK mostly for task scheduling I believe, as well as providing an object-based interface, so it's pretty tightly integrated, and probably can't be used in a non-gtk app.

Opened #330 re spell checking. I can't reproduce any issue with Ctrl-C / Ctrl-V; could be a system issue.

Thanks for all the feedback!

**Comment by szbnwer** [comment]
yes, that dev toolbar is what i'm using

i have to familiarize myself deeper with docs before ask anything :D anyway it would be good to gen docs in the repo or an another one like doing it automated on daily basis or whatever and point on it from the website, then even the wiki could fall out from the game and gives easier maintenance :)

< > works fine thx :)

i don't wanna use qt as well because tekui is a big love and as i will do my own stuffs therefore i will only pimp luakit from lua side...

only ctrl+c is not working, maybe i should try to make a syswide install instead of a local one... if that will help i will tell it...

ill try spellcheck in the rc, but if that's not the right place, then please tell me

currently i'm deeply in lack of time so i think the best would be to explore slowly things around instead of ask everything, but thx for all the solutions and hints and everything, so i'll be back when i'll have more time and find anything else - i think i already took much time from you as well instead of having my own research but so much thx for everything :)

**Comment by aidanholm** [comment]
The docs are auto-generated at the moment. I've been meaning to get them onto the website for a while now. The other problem is that the docs are still incomplete and need some work done on them.

The spellcheck API won't work yet; but I'll add that after the July release is done.

I'm always happy to help :) Let me know if you have any more questions!
